proposals/0118-partitioned-epoch-reward-distribution.md

```diff
@@ -0,0 +1,274 @@
+   for reward distribution. However, they also add addition cost to block
+   processing, especially for the first block in the epoch. The first block is
+   already computationally heavily - it is responsible for processing all the
+   reward computation. The additional cost of those sysvars puts more burden
+   onto that block and hurt the timing for it.
+
+## Detailed Design
+
+The major bottleneck for epoch reward distribution is to distribute rewards to
+stake accounts. At the time of writing, there are approximately 550K active
+stake accounts and 1.5K vote accounts on Solana Mainnet Beta. Given the
+relatively small number of vote accounts, it makes sense to keep vote rewards
+distribution mechanism unchanged. They can still be distributed efficiently at
+the first block of the epoch boundary. This reduces the impact of rewards for
+vote account and also simplifies the overall changes. It also lets us focus on
+solving the primary bottleneck - Stake Rewards. Only Stake rewards are going to
+be distributed out over multiple blocks.
+
+In the new stake rewards distribution approach, we will separate the
+computation of rewards from the actual distribution of rewards at the epoch
+boundary by dividing the process into two distinct phases:
+
+   1. rewards calculation phase - during which the epoch rewards for all
+      activate stake accounts are computed and distribution chunks are scheduled.
+
+   1. rewards distribution phase - during which the calculated epoch rewards
+      for the active stake accounts are distributed.
+
+To help maintain the total capital balance and track/verify the reward
+distribution during rewarding phases, a new sysvar account, `EpochRewards`, is
+proposed. The `EpochRewards` sysvar holds the balance of the rewards that are
+pending for distribution.
+
+
+### Rewards Calculation
+
+Reward calculation phase computes all the rewards that need to be distributed
+for the active stake accounts, and partitions the reward into a number of
+chunks for distribution in the next phase.
+
+Currently, on Solana Mainnet Beta with ~550K active stake accounts, epoch reward
+calculation takes around 10 seconds on average. This will make it impossible to
+perform rewards computation synchronous within one block.
+
+However, there are quite a few promising optimizations that can cut down the
+reward computation time. An experiment for reward calculation optimization
+(https://github.com/solana-labs/solana/pull/31193) showed that we can cut the
+reward calculation time plus vote reward distribution time to around 1s. This
+makes synchronous reward computation and asynchronous reward distribution a
+feasible approach. We also believe that there is room for more optimization to
+further cut down the above timing.
+
+Therefore, the following design is based on the above optimization. The reward
+calculation will be performed at the first block of the epoch, block height `X`.
```

Comment:
throughout we assume that the rewards distribution algorithm is known by the implementer. i'm not sure this is a strong assumption

Reply:
I assume you mean the rewards *calculation* algorithm?
I don't want to fully define that here, but if there's a good doc to link to, that would be good.

Reply:
Added this link